payments-core/src/main/java/com/stripe/android/IntentConfirmationInterceptor.kt

```diff
@@ -0,0 +1,226 @@
+            val paymentMethod = createPaymentMethod(paymentMethodCreateParams)
+
+            if (paymentMethod != null) {
+                intercept(
+                    clientSecret = null,
+                    paymentMethod = paymentMethod,
+                    shippingValues = shippingValues,
+                    setupForFutureUsage = setupForFutureUsage,
+                )
+            } else {
+                IntentConfirmationInterceptor.NextStep.Fail(
+                    error = IllegalStateException(
+                        "Failed to create payment method during confirmation"
+                    ),
+                )
+            }
+        }
+    }
+
+    override suspend fun intercept(
+        clientSecret: String?,
+        paymentMethod: PaymentMethod,
+        shippingValues: ConfirmPaymentIntentParams.Shipping?,
+        setupForFutureUsage: ConfirmPaymentIntentParams.SetupFutureUsage?,
+    ): IntentConfirmationInterceptor.NextStep {
+        return if (clientSecret != null) {
+            createConfirmStep(clientSecret, shippingValues, paymentMethod)
+        } else {
+            when (createIntentCallback) {
+                is CreateIntentCallbackForServerSideConfirmation -> {
+                    handleServerSideConfirmation(
+                        createIntentCallback = createIntentCallback,
+                        shouldSavePaymentMethod = setupForFutureUsage == OffSession,
+                        paymentMethod = paymentMethod,
+                        shippingValues = shippingValues
+                    )
+                }
+                is CreateIntentCallback -> {
+                    handleClientSideConfirmation(
+                        createIntentCallback = createIntentCallback,
+                        paymentMethod = paymentMethod,
+                        shippingValues = shippingValues
+                    )
+                }
+                else -> {
+                    IntentConfirmationInterceptor.NextStep.Fail(
+                        error = IllegalStateException(
+                            "ConfirmCallback must be implemented when using " +
+                                "IntentConfiguration to configure PaymentSheet",
+                        ),
+                    )
+                }
+            }
+        }
+    }
+
+    private suspend fun createPaymentMethod(params: PaymentMethodCreateParams): PaymentMethod? {
+        return runCatching {
+            stripeRepository.createPaymentMethod(
+                paymentMethodCreateParams = params,
+                options = requestOptions,
+            )
+        }.getOrNull()
+    }
+
+    private suspend fun handleClientSideConfirmation(
+        createIntentCallback: CreateIntentCallback,
+        paymentMethod: PaymentMethod,
+        shippingValues: ConfirmPaymentIntentParams.Shipping?
+    ): IntentConfirmationInterceptor.NextStep {
+        val result = createIntentCallback.onCreateIntent(paymentMethodId = paymentMethod.id!!)
```

Comment:
Is this a modeling problem? Why do we need to force unwrap the `paymentMethod.id`?

Reply:
Yes, but not unique to this project. It comes from the fact that `com.stripe.android.model.PaymentMethod` is modelled with a nullable ID. I’m not sure what the reason for this was.